src/scheduler/MpiThreadPool.cpp

```diff
@@ -0,0 +1,98 @@
+#include <faabric/scheduler/MpiThreadPool.h>
+#include <faabric/util/logging.h>
+
+namespace faabric::scheduler {
+MpiAsyncThreadPool::MpiAsyncThreadPool(int nThreads)
+  : shutdown(false)
+{
+    faabric::util::getLogger()->debug(
+      "Starting an MpiAsyncThreadPool of size {}", nThreads);
+
+    // Initialize job queue
+    localJobQueue = std::make_shared<MpiJobQueue>();
+
+    // Initialize thread pool
+    for (int i = 0; i < nThreads; ++i) {
+        threadPool.emplace_back(
+          std::bind(&MpiAsyncThreadPool::entrypoint, this, i));
+    }
+}
+
+MpiAsyncThreadPool::~MpiAsyncThreadPool()
+{
+
+    faabric::util::getLogger()->debug("Shutting down MpiAsyncThreadPool");
+    this->shutdown = true;
+
+    // When the destructor is called, either some thread will have hit the
+    // dequeue timeout and won't be joinable, or none has and all are joinable.
+    // We wait untill all hit the timeout, saving an additional cond. variable.
+    for (auto& thread : threadPool) {
+        if (thread.joinable()) {
+            thread.join();
+        }
+    }
+}
+
+void MpiAsyncThreadPool::awaitAsyncRequest(int reqId)
+{
+    faabric::util::UniqueLock lock(awakeMutex);
+
+    // Wait until the condition is signaled and the predicate is false.
+    // Note that the same lock protects concurrent accesses to asyncReqMap
+    awakeCV.wait(lock, [this, reqId] {
+        return this->asyncReqMap.find(reqId) != this->asyncReqMap.end();
```

Comment:
I wasn't aware of the predicate option in [`std::wait`](https://en.cppreference.com/w/cpp/thread/condition_variable/wait). Unless proven inefficient, it simplifies a lot the notify-all but wake-one logic.